dev/MANUALLY_GENERATING_IMAGE_CACHE_AND_CONSTRAINTS.md

```diff
@@ -532,3 +359,196 @@ breeze release-management update-constraints --constraints-repo /home/user/airfl
     --commit-message "Update pymssql constraint to 2.2.8 and Authlib to 1.3.0" \
     --airflow-constraints-mode constraints
 ```
+
+
+# Figuring out backtracking dependencies
+
+## Why we need to figure out backtracking dependencies
+
+Sometimes, very rarely the CI image in `canary` builds take a very long time to build. This is usually
+caused by `pip` trying to figure out the latest set of dependencies (`eager upgrade`) .
+The resolution of dependencies is a very complex problem and sometimes it takes a long time to figure out
+the best set of dependencies. This is especially true when we have a lot of dependencies and they all have
+to be found compatible with each other. In case new dependencies are released, sometimes `pip` enters
+a long loop trying to figure out if the newly released dependency can be used, but due to some other
+dependencies of ours it is impossible, but it will take `pip` a very long time to figure it out.
+
+This is visible in the "build output" as `pip` attempting to continuously backtrack and download many new
+versions of various dependencies, trying to find a good match.
+
+This is why we sometimes we need to help pip to skip newer versions of those dependencies, until the
+condition that caused the backtracking is solved.
+
+We do it by adding `dependency<=version` to the EAGER_UPGRADE_ADDITIONAL_REQUIREMENTS variable in
+`Dockerfile.ci`. The trick is to find the dependency that is causing the backtracking.
+
+Here is how. We use `bisecting` methodology to try out candidates for backtrack triggering among the
+candidates that have been released in PyPI since the last time we successfully run
+``--upgrade-to-newer-dependencies`` and committed the constraints in the `canary` build.
+
+## How to figure out backtracking dependencies
+
+First - we have a breeze command that can help us with that:
+
+```bash
+breeze ci find-backtracking-candidates
+```
+
+This command should be run rather quickly after we notice that the CI build is taking a long time and fail,
+because it is based on the fact that eager upgrade produced valid constraints at some point of time and
+it tries to find out what dependencies have been added since then and limit them to the version that
+was used in the constraints.
```

Comment:
```suggestion
This command should be run quickly after noticing that the CI build is taking a long time and failing. It relies on the fact that an eager upgrade produced valid constraints at some point of time, so it tries to identify which dependencies have been added since then. By doing it, it limits them to the versions used in the constraints.
```